### PR TITLE
Create public/private functions for crm_resource -D

### PR DESCRIPTION
**Pull request by clumens**

### CODE REVIEW

File: include/pacemaker.h [hunk 1]
```diff
@@ -195,6 +195,19 @@ pcmk_query_node_name(xmlNodePtr *xml, uint32_t node_id, char **node_name,
...
+/*!
+ * \brief Remove a resource
+ *
+ * \param[in] rsc_id    Resource to remove
+ * \param[in] rsc_type  Type of the resource ("primitive", "group", etc.)
+ *
+ * \return Standard Pacemaker return code
+ * \note Due to how the underlying CIB functions work, this function will
+ *       return pcmk_rc_ok even if \p rsc_type is incorrect for \p rsc_id,
+ *       or \p rsc_id doesn't exist at all.
```
Comment (kgaillot) [inline include/pacemaker.h:207]: FYI, that is the desired behavior, since deleting something that doesn't exist is always successful. That's perhaps questionable if rsc_id exists with a different rsc_type, but the requested object technically doesn't exist.

File: lib/pacemaker/pcmk_resource.c [hunk 2]
```diff
@@ -103,6 +104,68 @@ best_op(const pcmk_resource_t *rsc, const pcmk_node_t *node)
...
+ */
+int
+pcmk__resource_delete(cib_t *cib, uint32_t cib_opts, const char *rsc_id,
+                      const char *rsc_type)
+{
+    int rc = pcmk_rc_ok;
+    xmlNode *msg_data = NULL;
+
+    if (cib == NULL) {
+        return ENOTCONN;
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_resource.c:126]: FYI we could open and close a local connection within this function (like the public one) when cib==NULL, but until we have a use case there's no need

**Review by kgaillot — COMMENTED** [review]
Nice!

File: lib/pacemaker/tests/pcmk_resource/pcmk_resource_delete_test.c [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    cib_t *cib = NULL;
+
+    cib = cib_new();
+    cib->cmds->signon(cib, crm_system_name, cib_command);
+
+    pcmk__g_strcat(xpath,
+                   pcmk_cib_xpath_for(XML_CIB_TAG_RESOURCES),
+                   "//*[@" XML_ATTR_ID "=\"", rsc, "\"]", NULL);
+
+    fprintf(stderr, "%s\n", xpath->str);
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/tests/pcmk_resource/pcmk_resource_delete_test.c:108]: Hah, yes.  Removed.

File: include/pacemaker.h (hunk 1 above)
Comment (clumens) [reply include/pacemaker.h:207]: I think it's worth leaving a comment in place just to give people a heads up, but I've changed it to make it sound less like a bug.

File: lib/pacemaker/pcmk_resource.c (hunk 2 above)
Comment (clumens) [reply lib/pacemaker/pcmk_resource.c:126]: At the moment, the private functions are a delightful mix of expecting an already open cib connection (pcmk__status) and creating their own connection without taking it as a parameter (pcmk__simulate).  I can do whatever here, but I feel like we should standardize the behavior across all existing private functions so we have less to think about later.

File: lib/pacemaker/pcmk_resource.c [hunk 4]
```diff
@@ -103,6 +104,68 @@ best_op(const pcmk_resource_t *rsc, const pcmk_node_t *node)
...
+
+    rc = cib->cmds->remove(cib, XML_CIB_TAG_RESOURCES, msg_data, cib_opts);
+    rc = pcmk_legacy2rc(rc);
+
+    free_xml(msg_data);
+    return rc;
+}
+
+int
+pcmk_resource_delete(const char *rsc_id, const char *rsc_type)
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_resource.c:144]: A couple things to note on this:

1. I can't find any instance of a public libpacemaker function handling an xmlNodePtr argument without also using pcmk__xml_output_new (and therefore, a pcmk__output_t) at the same time.  Deleting doesn't do any output, so this is a little weird, but whatever.  I just think it's interesting this hasn't come up before.
2. Several of the public pcmk_fence_* functions do not take an xmlNodePtr - seems like we need to clean that up.
3. I don't think we even handle the possibility of failure right now, unless somewhere along the lines, something outputs an error.  For instance, look at pcmk_resource_digests.  It calls pcmk__resource_digests, which can return various error codes.  Those error codes will be returned by the public function, but nowhere will they end up in the XML node because pcmk__xml_output_finish doesn't take the error code as a parameter.  It previously did, but it also used that to decide whether to do any output at all.  Maybe it makes sense to take the error code and add it (with out->error?).

File: lib/pacemaker/pcmk_resource.c (hunk 2 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_resource.c:126]: If we want common behavior, then passing the cib connection and opening a local one if it's NULL makes the most sense. I wouldn't worry about it with this PR though, we can add it separately

File: lib/pacemaker/pcmk_resource.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_resource.c:144]: Yep, we should definitely create an XML output object for all public APIs. Even if there is no output, the XML will have the exit status code and message. It lets users have consistent handling for all calls. The XML definitely has a `<status code="N" message="BLAH">` element; I'm not sure whether that's being set somewhere else or we're always returning success for digests. At least crm_attribute can return an error code there.

Delete probably *should* have success/failure output. Maybe for 3.0.0. 

The fence APIs aren't really public, they're guarded by an undefined constant. `stonith_t` will eventually be replaced by`pcmk_ipc_api_t`, so I don't want to expose anything new using it. When we get to that point, we can either make the fence APIs take a `pcmk_ipc_api_t`, or just create it within each function like we do with other connections.

[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_resource.c:144]: > Yep, we should definitely create an XML output object for all public APIs. Even if there is no output, the XML will have the exit status code and message. It lets users have consistent handling for all calls. The XML definitely has a `<status code="N" message="BLAH">` element; I'm not sure whether that's being set somewhere else or we're always returning success for digests. At least crm_attribute can return an error code there.

Right, I think this is the difference between using the public and private functions.

In a tool like crm_attribute, we are using the private functions (or, we would be if we're using any at all - I haven't checked).  We create a pcmk__output_t, call some function, and get a return value.  At the end, we call `out->finish()` with that return value.  It's `xml_finish` that constructs the wrapper around the rest of the XML with the return code and message.

If we're using the public functions, we do not have a pcmk__output_t to begin with.  The public function creates one, calls a function, and gets a return value.  We call `pcmk__xml_output_finish` which doesn't take a return code argument, which then calls `out->finish()` and only ever passes 0 for the return code.  So, public function XML will always be wrapped up with success.  The return code gets passed back to the caller of the public function which can still exit with error, but the output has already happened.

Long story short, I think `pcmk__xml_output_finish` needs to take a return code and pass that through to `out->finish()`.

[the remark this answers is not in the document]
Comment (clumens) [reply lib/pacemaker/pcmk_resource.c:144]: See #3270.

**Comment by clumens** [comment]
I think there's nothing left to do here aside from the NULL cib connection stuff which I'll look at separately.  I've rebased this against the latest main.

File: lib/pacemaker/pcmk_resource.c [hunk 5]
```diff
@@ -228,6 +228,7 @@ pcmk__resource_digests(pcmk__output_t *out, pcmk_resource_t *rsc,
     return rc;
 }
 
+// @COMPAT The scheduler parameter is unused and can be removed at the next break
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_resource.c:231]: Yeah we should have designed this one better all around. As a high-level function, it should probably take the resource and node names instead of objects, and connect to the CIB to get the objects. Unrelated to this PR though.

**Review by kgaillot — APPROVED** [review]
